worker/peergrouper/worker_test.go

```diff
@@ -649,44 +615,133 @@ func (s *workerSuite) runUntilPublish(c *gc.C, st *fakeState, errMsg string) {
+	// Replica set addresses should change to the new space.
+	st.setHASpace("three")
+	mustNext(c, memberWatcher, "waiting for members to be updated for space change")
+	assertMemberAddresses(c, st, ipVersion.formatHost, 3)
+}
+
+// All machines have the same address in this test for simplicity.
+// The space three address is 0.0.0.3 giving us the host port of 0.0.0.3:4711.
+func assertMemberAddresses(c *gc.C, st *fakeState, addrTemplate string, addrDesignator int) {
+	members, _ := st.session.CurrentMembers()
+	c.Assert(members, gc.HasLen, 3)
+	for i := 0; i < 3; i++ {
+		c.Assert(members[i].Address, gc.Equals,
+			net.JoinHostPort(fmt.Sprintf(addrTemplate, addrDesignator), fmt.Sprint(mongoPort)))
+	}
+}
+
+func (s *workerSuite) TestReturnsErrorWhenNoHASpaceAndMachinesWithMultiLocalAddr(c *gc.C) {
 	DoTestForIPv4AndIPv6(c, s, func(ipVersion TestIPVersion) {
-		publishCh := make(chan [][]network.HostPort, 100)
+		st, machines, addrs := mongoSpaceTestCommonSetup(c, ipVersion, false)
 
-		count := 0
-		publish := func(apiServers [][]network.HostPort) error {
-			c.Logf("** publish call: count=%d", count)
-			publishCh <- apiServers
-			count++
-			if count <= 3 {
-				return fmt.Errorf("publish error")
-			}
-			return nil
+		for i, machine := range machines {
+			// machine 10 gets an address in space one
+			// machine 11 gets addresses in spaces one and two
+			// machine 12 gets addresses in spaces one, two and three
+			st.machine(machine).setAddresses(addrs[:i+1]...)
 		}
-		st := NewFakeState()
-		InitState(c, st, 3, ipVersion)
 
-		w := s.newWorker(c, st, st.session, SetAPIHostPortsFunc(publish))
-		defer workertest.CleanKill(c, w)
+		err := s.newWorker(c, st, st.session, nopAPIHostPortsSetter{}).Wait()
+		errMsg := `computing desired peer group: machine "1[12]" has more than one non-local address and juju-ha-space is not set`
+		c.Check(err, gc.ErrorMatches, errMsg)
+	})
+}
 
-		retryInterval := initialRetryInterval
-		for i := 0; i < 4; i++ {
-			s.clock.WaitAdvance(retryInterval, coretesting.ShortWait, 1)
-			retryInterval = scaleRetry(retryInterval)
-			select {
-			case servers := <-publishCh:
-				AssertAPIHostPorts(c, servers, ExpectedAPIHostPorts(3, ipVersion))
-			case <-time.After(coretesting.LongWait):
-				c.Fatalf("timed out waiting for publish #%d", i)
-			}
+func (s *workerSuite) TestWorkerRetriesOnSetAPIHostPortsErrorIPv4(c *gc.C) {
+	s.doTestWorkerRetriesOnSetAPIHostPortsError(c, testIPv4)
+}
+
+func (s *workerSuite) TestWorkerRetriesOnSetAPIHostPortsErrorIPv6(c *gc.C) {
+	s.doTestWorkerRetriesOnSetAPIHostPortsError(c, testIPv6)
+}
+
+func (s *workerSuite) doTestWorkerRetriesOnSetAPIHostPortsError(c *gc.C, ipVersion TestIPVersion) {
+	logger.SetLogLevel(loggo.TRACE)
+
+	publishCh := make(chan [][]network.HostPort, 100)
+	failedOnce := false
+	publish := func(apiServers [][]network.HostPort) error {
+		if !failedOnce {
+			failedOnce = true
+			return fmt.Errorf("publish error")
 		}
+		publishCh <- apiServers
```

Comment:
general guideline is to not allow bare sends/receives on channels, but always select{} with an ultimate timeout so you never hang the test.
I guess here you're expecting that a buffer of 100 is guaranteed big enough?

Reply:
Buffer was there prior, but the test has actually changed to expect just one publish event instead of publishing every time even when an error was returned. I knocked a zero off the buffer.

This one actually became more fiddly to test, because there is a path where the controller topology is published first, with setAPIHostPorts throwing an error, but then the config change wakes up the loop and runs through setting the retry interval back to default.